File: Data_structure/Fenwick_Tree/FenwickTree.kt
/**
 * 1-based 펜윅 트리
 */
class FenwickTree(
    val arr: Array<Int>,
) {
    val n: Int
        get() = arr.size

    private val tree = Array(n + 1) { 0 }

    private fun lsb(idx: Int) = idx and -idx

    init {
        build()
    }

    fun update(index: Int, value: Int) {
        require(index in 0 until n) { "index 범위 초과: index=$index, n=$n" }
        updateDiff(index, value - arr[index])
        arr[index] = value
    }

    fun rangeSum(left: Int, right: Int): Int {
        require(left <= right) { "left는 right보다 클 수 없습니다. left=$left, right=$right" }
        // 1-base 라서 각자 +1
        return prefixSum(right + 1) - prefixSum(left)
    }

    private fun prefixSum(
        index: Int
    ): Int {
        var sum = 0
        var idx = index
        while (idx > 0) {
            // 비트 연상으로 구한다
            sum += tree[idx]
            idx -= lsb(idx)
        }
        return sum
    }

    private fun updateDiff(index: Int, diff: Int) {
        // 펜윅 트리는 1-based 라서 1을 더한다
        var idx = index + 1
        while (idx <= n) {
            tree[idx] += diff
            idx += lsb(idx)      // i가 담당하는 구간 길이만큼 점프
        }
    }

    private fun build() {
        arr.indices.forEach { i ->
            updateDiff(i, arr[i])
        }
    }
}

File: Data_structure/Fenwick_Tree/FenwickTree.py
from typing import List


class FenwickTree:
    """
    1-based FenwickTree
    """

    def __init__(
            self,
            arr: List[int]
    ):
        self.size = len(arr)
        self.arr = arr
        self.tree = [0] * (len(arr) + 1)
        self.__build()

    def range_sum(
            self,
            left: int,
            right: int
    ) -> int:
        """
        index left~right 까지의 합

        :param left: 범위의 시작 index
        :param right: 범위위 종료 index

        :return: left ~ right 까지의 합
        """
        assert left <= right
        return self.__prefix_sum(right + 1) - self.__prefix_sum(left)

    def update(
            self,
            index: int,
            value: int
    ):
        """
        배열의 index의 값을 value로 업데이트

        :param index: 업데이트할 인덱스
        :param value: 해당 배열의 값
        """
        assert index in range(0, self.size)
        self.__update_diff(index, value - self.arr[index])
        self.arr[index] = value

    def __lsb(self, idx) -> int:
        """
        :return: 해당 값의 가장 오른쪽 1비트
        """
        return idx & -idx

    def __build(self):
        """
        초기 값을 세팅
        """
        for item in range(0, self.size):
            self.__update_diff(item, self.arr[item])

    def __update_diff(
            self,
            index: int,
            diff: int
    ) -> None:
        """
        해당 index의 구간 합을 가진 index들을 업데이트
        :param index: 현재 업데이트 할
        :param diff: 업데이트 될 값의 차이
        """
        idx = index + 1
        while idx <= self.size:
            self.tree[idx] += diff
            idx += self.__lsb(idx)

    def __prefix_sum(
            self,
            index: int
    ) -> int:
        """
        해당 index까지의 LSB를 이용해 구간 합을 구한다.
        """
        prefix_sum = 0
        idx = index
        while idx > 0:
            prefix_sum += self.tree[idx]
            idx -= self.__lsb(idx)
        return prefix_sum


if __name__ == "__main__":
    fenwick_tree = FenwickTree(
        [1, 2, 3, 4, 5]
    )
    print(
        fenwick_tree.range_sum(0, 4)
    )
